File: notebooks/12-array_to_bst.py
from __future__ import annotations


class TreeNode:
    def __init__(self, val: int = 0, left: TreeNode | None = None, right: TreeNode | None = None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sortedArrayToBST(self, nums: list[int]) -> TreeNode | None:
        if not nums:
            return None
        mid = len(nums) // 2
        node = TreeNode(nums[mid])
        node.left  = self.sortedArrayToBST(nums[:mid])
        node.right = self.sortedArrayToBST(nums[mid + 1:])
        return node


def is_valid_bst(node: TreeNode | None, lo: float = float('-inf'), hi: float = float('inf')) -> bool:
    if node is None:
        return True
    if not (lo < node.val < hi):
        return False
    return is_valid_bst(node.left, lo, node.val) and is_valid_bst(node.right, node.val, hi)


def height(node: TreeNode | None) -> int:
    if node is None:
        return 0
    return 1 + max(height(node.left), height(node.right))


def is_height_balanced(node: TreeNode | None) -> bool:
    if node is None:
        return True
    return (abs(height(node.left) - height(node.right)) <= 1
            and is_height_balanced(node.left)
            and is_height_balanced(node.right))


def inorder(node: TreeNode | None) -> list[int]:
    if node is None:
        return []
    return inorder(node.left) + [node.val] + inorder(node.right)


if __name__ == "__main__":
    s = Solution()

    # 用例 1: 空数组 → None
    assert s.sortedArrayToBST([]) is None

    # 用例 2: 单元素
    t = s.sortedArrayToBST([1])
    assert t.val == 1 and t.left is None and t.right is None

    # 用例 3: 奇数长度 [-10, -3, 0, 5, 9]
    t = s.sortedArrayToBST([-10, -3, 0, 5, 9])
    assert inorder(t) == [-10, -3, 0, 5, 9]
    assert is_valid_bst(t)
    assert is_height_balanced(t)

    # 用例 4: 偶数长度 [1, 2, 3, 4]
    t = s.sortedArrayToBST([1, 2, 3, 4])
    assert inorder(t) == [1, 2, 3, 4]
    assert is_valid_bst(t)
    assert is_height_balanced(t)

    # 用例 5: 较长数组，验证高度平衡
    t = s.sortedArrayToBST(list(range(1, 16)))
    assert inorder(t) == list(range(1, 16))
    assert is_valid_bst(t)
    assert is_height_balanced(t)

    print("所有测试用例通过 ✅")

File: notebooksc++/12-array_to_bst.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <algorithm>

using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int val = 0, TreeNode* left = nullptr, TreeNode* right = nullptr)
        : val(val), left(left), right(right) {}
};

TreeNode* buildBST(const vector<int>& nums, int lo, int hi) {
    if (lo > hi) return nullptr;
    int mid = lo + (hi - lo) / 2;
    TreeNode* node = new TreeNode(nums[mid]);
    node->left = buildBST(nums, lo, mid - 1);
    node->right = buildBST(nums, mid + 1, hi);
    return node;
}

TreeNode* sortedArrayToBST(const vector<int>& nums) {
    return buildBST(nums, 0, static_cast<int>(nums.size()) - 1);
}

bool isValidBST(TreeNode* node, long lo, long hi) {
    if (node == nullptr) return true;
    if (node->val <= lo || node->val >= hi) return false;
    return isValidBST(node->left, lo, node->val) && isValidBST(node->right, node->val, hi);
}

int height(TreeNode* node) {
    if (node == nullptr) return 0;
    return 1 + max(height(node->left), height(node->right));
}

bool isHeightBalanced(TreeNode* node) {
    if (node == nullptr) return true;
    return abs(height(node->left) - height(node->right)) <= 1
        && isHeightBalanced(node->left)
        && isHeightBalanced(node->right);
}

vector<int> inorder(TreeNode* node) {
    if (node == nullptr) return {};
    vector<int> result;
    vector<int> leftVals = inorder(node->left);
    result.insert(result.end(), leftVals.begin(), leftVals.end());
    result.push_back(node->val);
    vector<int> rightVals = inorder(node->right);
    result.insert(result.end(), rightVals.begin(), rightVals.end());
    return result;
}

void deleteTree(TreeNode* node) {
    if (node == nullptr) return;
    deleteTree(node->left);
    deleteTree(node->right);
    delete node;
}

int main() {
    TreeNode* t = sortedArrayToBST({});
    assert(t == nullptr);

    t = sortedArrayToBST({1});
    assert(t->val == 1 && t->left == nullptr && t->right == nullptr);
    deleteTree(t);

    t = sortedArrayToBST({-10, -3, 0, 5, 9});
    assert(inorder(t) == vector<int>({-10, -3, 0, 5, 9}));
    assert(isValidBST(t, numeric_limits<long>::min(), numeric_limits<long>::max()));
    assert(isHeightBalanced(t));
    deleteTree(t);

    t = sortedArrayToBST({1, 2, 3, 4});
    assert(inorder(t) == vector<int>({1, 2, 3, 4}));
    assert(isValidBST(t, numeric_limits<long>::min(), numeric_limits<long>::max()));
    assert(isHeightBalanced(t));
    deleteTree(t);

    vector<int> nums(15);
    for (int i = 0; i < 15; i++) {
        nums[i] = i + 1;
    }
    t = sortedArrayToBST(nums);
    vector<int> expected(15);
    for (int i = 0; i < 15; i++) {
        expected[i] = i + 1;
    }
    assert(inorder(t) == expected);
    assert(isValidBST(t, numeric_limits<long>::min(), numeric_limits<long>::max()));
    assert(isHeightBalanced(t));
    deleteTree(t);

    cout << "All test cases passed ✅" << endl;
    return 0;
}
